numpy/core/src/multiarray/scalarapi.c

```diff
@@ -314,54 +322,48 @@ PyArray_FromScalar(PyObject *scalar, PyArray_Descr *outcode)
                 NULL, (PyObject *)scalar);
     }
 
-    PyArrayObject *r = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type,
-            typecode,
-            0, NULL,
-            NULL, NULL, 0, NULL);
+    if (outcode == NULL) {
+        outcode = typecode;
+        Py_INCREF(outcode);
+    }
+
+    PyArrayObject *r = (PyArrayObject *)PyArray_NewFromDescr_int(
+            &PyArray_Type, outcode, 0, NULL, NULL, NULL,
+            0, NULL, NULL, 0, 1);
     if (r == NULL) {
-        Py_XDECREF(outcode);
         return NULL;
     }
-    /* the dtype used by the array may be different to the one requested */
-    typecode = PyArray_DESCR(r);
-    if (PyDataType_FLAGCHK(typecode, NPY_USE_SETITEM)) {
-        if (typecode->f->setitem(scalar, PyArray_DATA(r), r) < 0) {
+    if (PyArray_DESCR(r) != outcode) {
+        /*
+         * Sanity check, only subarray dtypes should be modified, and there
+         * should be no subarray scalars (they would be arrays).
+         */
+        PyErr_SetString(PyExc_RuntimeError,
+                "dtype changed while creating an array from a scalar. This "
+                "should not be possible. Please notify the NumPy developers "
+                "you see this message");
```

Comment:
May as well make this message contain information we might care about:
```suggestion
        PyErr_Format(PyExc_RuntimeError,
                "dtype changed from %R to %R while creating an array from a scalar. This "
                "should not be possible. Please notify the NumPy developers "
                "you see this message", outcode, PyArray_DESCR(r));
```